src/java.base/unix/classes/sun/nio/fs/UnixPath.java

```diff
@@ -863,7 +867,40 @@ public Path toRealPath(LinkOption... options) throws IOException {
         } catch (UnixException x) {
             x.rethrowAsIOException(result);
         }
-        return result;
+
+        // Return if the file system is not both case insensitive and retentive
+        if (!fs.isCaseInsensitiveAndRetentive())
+            return result;
+
+        UnixPath path = fs.rootDirectory();
+
+        // Traverse the result obtained above from the root downward, leaving
+        // any '..' elements intact, and replacing other elements with the
+        // entry in the same directory to which it is equal ignoring case
+        for (int i = 0; i < result.getNameCount(); i++ ) {
+            UnixPath elt = result.getName(i);
+            String els = elt.toString();
+
+            // If the element is "..", append it directly and continue
+            if (els.equals("..")) {
+                path = path.resolve(elt);
+                continue;
+            }
+
+            // Obtain the stream of entries in the directory corresponding
+            // to the path constructed thus far, and extract the entry whose
+            // name is equal ignoring case to the name of the current element
+            try (Stream<Path> entries = Files.list(path)) {
```

Comment:
One initial comment on this is that you can't use Files.list in the provider implementation because the default provider can be configured to be something else. Also the security manager is still a supported execution mode so we'll need to work through the issues too.

Reply:
I think you'll have to change this to use UnixDirectoryStream here, it shouldn't be too hard. We'll also need to study the case check as I suspect that will need macOS specific code.

Reply:
Resolved (I think) in b6170562187932311597047524647eec66ad006f.